File: backend/models/payment.go
package models

import "gorm.io/gorm"

type Payment struct {
	gorm.Model

	CustomerID *uint `gorm:"not null"`
	CreatedByUserID uint `gorm:"not null"`
	TotalAmount float64 `gorm:"not null"`

	Customer *Customer `gorm:"foreignKey:CustomerID"`
	CreatedByUser User `gorm:"foreignKey:CreatedByUserID"`
	Allocations []PaymentAllocation `gorm:"foreignKey:PaymentID"`
}

type PaymentAllocation struct {
	gorm.Model

	PaymentID uint `gorm:"not null"`
	InvoiceID uint `gorm:"not null"`
	Amount float64 `gorm:"not null"`

	Payment Payment `gorm:"foreignKey:PaymentID"`
	Invoice Invoice `gorm:"foreignKey:InvoiceID"`
}

File: frontend/src/types/payment.ts
export interface PaymentCustomer {
  id: number;
  name: string;
  phone: string;
  isActive: boolean;
}

export interface PaymentUser {
  id: number;
  username: string;
}

export interface PaymentAllocationInvoice {
  id: number;
  totalAmount: number;
  paidAmount: number;
  status: string;
}

export interface PaymentAllocation {
  id: number;
  invoiceID: number;
  amount: number;
  invoice: PaymentAllocationInvoice;
}

export interface Payment {
  id: number;
  customerID?: number | null;
  customer?: PaymentCustomer | null;
  createdByUserID: number;
  createdByUser?: PaymentUser | null;
  totalAmount: number;
  createdAt: string;
  allocations: PaymentAllocation[];
}

export interface CreatePaymentAllocationPayload {
  invoiceID: number;
  amount: number;
}

export interface CreatePaymentPayload {
  customerID: number;
  totalAmount: number;
  allocations: CreatePaymentAllocationPayload[];
}

export interface CreatePaymentResponse {
  message?: string;
  data: Payment;
}

export interface PaymentDetailResponse {
  message?: string;
  data: Payment;
}

export interface PaginatedPayments {
  data: Payment[];
  page: number;
  limit: number;
  total: number;
  totalPages: number;
}

export interface PaymentListParams {
  page?: number;
  limit?: number;
  customerID?: number;
  fromDate?: string;
  toDate?: string;
}
